### PR TITLE
Add support to traverse all python collection objects

### LINKED ISSUE #752: Traversal of datapipe graphs broken for pipelines containing SampleMultiplexerDataPipe
## Issue description

Sharding as well as shuffling an iterable datapipe does not work if the datapipe contains a SampleMultiplexerDataPipe at any point after the ShardingFilter.
The root cause for this is that the graph traversal (torch.utils.data.graph.traverse) called by the DataLoader during apply_sharding() cannot traverse past a SampleMultiplexerDataPipe. The reason for this is that during graph traversal getstate_hook() tests whether any of the elements of the object state is an instance of IterDataPipe, MapDataPipe or tuple. In the case of SampleMultiplexerDataPipe the connected pipelines are, however, stored in a dictionary. As a consequence the graph traversal stops at this point and apply_sharding() and apply_shuffling() cannot be called on the relevant elements of the pipeline.

## Code example

    from torch.utils.data import DataLoader
    from torchdata.datapipes.iter import IterableWrapper, SampleMultiplexer

    pipe1 = IterableWrapper(range(2))
    pipe1 = pipe1.sharding_filter()
    pipe1 = pipe1.shuffle()

    pipe2 = IterableWrapper(range(2, 4))
    pipe2 = pipe2.sharding_filter()
    pipe2 = pipe2.shuffle()

    pipe = SampleMultiplexer({pipe1: 0.5, pipe2: 0.5})

    data_loader = DataLoader(pipe, shuffle=True, num_workers=2)

    for sample in data_loader:
        print(sample)

Running this code produces the following warning, even though the pipelines contain two Shufflers:

    UserWarning: `shuffle=True` was set, but the datapipe does not contain a `Shuffler`. Adding one at the end. Be aware that the default buffer size might not be sufficient for your task.

This indicates that Shuffle was not found during graph traversal.
Furthermore, each element produced by the dataset is emitted n times, where n = num_workers. This indicates that sharding is not working.

Expected output:
- no warning about missing Shuffle
- tensors with the values 0, 1, 2, and 3 in an arbitrary order with each number printed exactly once

## System Info
PyTorch version: 1.12.1+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Gentoo Base System release 2.8 (x86_64)
GCC version: (Gentoo 11.3.0 p4) 11.3.0
Clang version: 14.0.6
CMake version: version 3.22.4
Libc version: glibc-2.35

Python version: 3.10.5 (main, Jun 23 2022, 09:37:51) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.59-gentoo-x86_64-Intel-R-_Xeon-R-_E-2146G_CPU_@_3.50GHz-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.124
GPU models and configuration: 
GPU 0: NVIDIA GeForce GTX 1050 Ti
GPU 1: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 510.85.02
cuDNN version: Probably one of the following:
/opt/cuda/targets/x86_64-linux/lib/libcudnn.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] torch==1.12.1+cu116
[pip3] torchdata==0.4.1
[pip3] torchvision==0.13.1+cu116
[conda] Could not collect

Process finished with exit code 0

cc @SsnL @VitalyFedyunin @ejguan @NivekT

### PR DESCRIPTION
**Pull request by ejguan**
Fixes https://github.com/pytorch/data/issues/752

This PR makes `traverse` function supporting more collections data structures from Python. The `getstate_hook` will be invoked after custom `__getstate__` function. This would guarantee that `traverse` function will be working as long as the `DataPipe` is working properly with multiprocessing.

### CODE REVIEW

**Review by NivekT — COMMENTED** [review]
Have you considered this instead? Or is this too broad?
```python
from collections.abc import Collection

isinstance(obj, Collection)  # Returns True for list, tuple, dict, set
```
The practical difference is that the above will cover any `Sequence` and also `ValuesView`. 

https://docs.python.org/3/library/collections.abc.html

**Comment by ejguan** [comment]
> Have you considered this instead? Or is this too broad?
> 
> ```python
> from collections.abc import Collection
> 
> isinstance(obj, Collection)  # Returns True for list, tuple, dict, set
> ```
> 
> The practical difference is that the above will cover any `Sequence` and also `ValuesView`.
> 
> https://docs.python.org/3/library/collections.abc.html

Yeah. I did think about using `Iterable` from the `collections.abc`, which also covers all `list`, `tuple`, ... objects. And, I was thinking it would be too broad.

`Collection` seems like a better idea than `Iterable`. Thanks for your suggestion and I will change it.

**Review by NivekT — APPROVED** [review]
LGTM!

**Comment by weiwangmeta** [comment]
@pytorchbot revert -m="Diff reverted internally" -c="ghfirst"

Will re-merge this after smoothing out sync issues.

**Comment by NivekT** [comment]
I think this was merged into nightly and broke some TorchData's TorchVision related CI tests.

https://github.com/pytorch/data/runs/8074482191?check_suite_focus=true#step:9:2359

Let's not land this for now.

cc: @VitalyFedyunin @ejguan

**Comment by ejguan** [comment]
It turns out that `list` is the data type that potentially violates `traverse` function. The failure we have encountered in TorchVision is the buffer from `Demux` contains a list of file streams that can not be `pickled`.

**Review by NivekT — COMMENTED** [review]
If it can happen to `list`, then I believe it can happen to any `Collection` since the object might contain unpickable streams.

Is it possible to always call `traverse` before the buffers of DataPipes are populated?
- It will require users to always call `apply_sharding`, `apply_shuffle_seed`, `apply_shuffle_settings` and etc before fill the DataPipes' buffer (through iteration or restoration).
- One thought came to mind is that we can save a copy of the graph in the beginning (maybe after applying adapter functions) and re-use that copy of the graph, but that would require no further insertion/removal of DataPipes. Not sure how feasible that is.

**Comment by ejguan** [comment]
> Is it possible to always call `traverse` before the buffers of DataPipes are populated?

It might not be possible for torchvision's use case as they are loading some resources during the construction of DataPipe. I do have a workaround and will update the PR to see your opinion.

**Comment by ejguan** [comment]
> I do have a workaround

The followings are two options in my mind.

1. Make collections of `DataPipes` with special classes. TorchData provides special form of `list`, `dict`, `tuple` that hold `DataPipes`. Then, `traverse` function would only take `DataPipe` or those special collection objects. It's easier for us to extract graph even during the iteration but it adds extra burden for DataPipe developers.
2. Convert `getstate_hook` to `getstate_wrapper`, which will rely on the implementation of `__getstate__` function from each `DataPipe`. `traverse` will only go through `DataPipe` or common python collection objects among the returned state from the original `__getstate__` function. The difference becomes the original `__getstate__` function should be able to handle unserialized buffer. Like `demux`, it currently ignores `buffer` during serialization and always recreate `buffer` during deserialization. However, the same problem still exists that `traverse` function might not be able to be called after `DataPipe`'s iteration.

**Comment by ejguan** [comment]
Added a PR to guarantee this change won't break TorchVision: https://github.com/pytorch/vision/pull/6569

But, IMO, this **CAN NOT** be the ultimate solution for TorchData since we would require users to always do lazy loading even with their custom data pipeline. A proper solution will be updated in this PR.

**Comment by ejguan** [comment]
Summarize the offline discussion with @VitalyFedyunin 

We agree on a comment that if the graph of DataPipe is working with Multiprocessing, it should also works with `traverse` function. So, our decision is to still rely on `hook`. But instead of calling hook before the custom `__getstate__` function, we will invoke hook directly over the `state` returned by `__getstate__` function. (using `self.__dict__` by default)

And, for future proofing, we might want to rely on wrapper over `__iter__` of `DataPipe` to always use a copy of `self` when it's not within a context of `DataLoader` in a single-process `DataLoader2`.

File: torch/utils/data/graph.py
Comment (NivekT) [inline torch/utils/data/graph.py:89]: nit: it seems like the default usages always set `only_datapipe=True`? Can you provide an example when you want it to be `False`? And perhaps mention that example in here for future reference?

File: torch/utils/data/graph.py
Comment (NivekT) [inline torch/utils/data/graph.py:48]: Question: Why is this removed? To allow visiting the same obj multiple times?

**Review by NivekT — COMMENTED** [review]
Question: Does this change allow buffers to be non-empty when traversing the graph? If so, how?

**Comment by NivekT** [comment]
> And, for future proofing, we might want to rely on wrapper over `__iter__` of `DataPipe` to always use a copy of `self` when it's not within a context of `DataLoader` in a single-process `DataLoader2`.

Sorry I missed the meeting, but what does this mean? We should create an isolated copy of the DataPipe when `__iter__` is called? So that the state is independent from the DataPipe object itself?

**Comment by ejguan** [comment]
> Does this change allow buffers to be non-empty when traversing the graph? If so, how?

No, it won't. Rather than let `traverse` function handle non-empty buffer itself, we decided to make it relying on the implementation of `IterDataPipe.__getstate__`. All non-empty buffer should be specially handled within `__getstate__` function, as the same problem would happen for multiprocessing.

> Sorry I missed the meeting, but what does this mean? We should create an isolated copy of the DataPipe when `__iter__` is called? So that the state is independent from the DataPipe object itself?

As we have discussed before, when `DataPipe` is iterated by `DataLoader2`, it should be a copy of the original `DataPipe` graph being iterated. Then, for the case that `DataPipe` is iterated without using `DataLoader2`, to guarantee that the original `DataPipe` graph won't generate any `buffer`, a new copy of `DataPipe` graph should be used for the iteration of `DataPipe`.

File: torch/utils/data/graph.py
Comment (ejguan) [reply torch/utils/data/graph.py:89]: This is for BC. We previously don't have `only_datapipe` option for `traverse` function. So, previously, this function behaves like `only_datapipe=False`.
I can add a deprecation warning here to indicate users to remove `only_datapipe` argument.

Comment (NivekT) [reply torch/utils/data/graph.py:89]: Yes I think a deprecation warning will be good if we do not expect it to be used by users in the future.

**Comment by ejguan** [comment]
> The reasoning makes sense but it might get tricky for `ChildDataPipe` cases. I'd imagine you want to stay on the same graph while iterating through two `ChildDataPipes` with the same parent. There might be a trivial solution but I will need to think more.

We might need to do another follow-up with Vitaly on this topic.

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**
